ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java

```diff
@@ -0,0 +1,1760 @@
+      throws IOException, StoreException {
+    // NOTE: This function fails on a corner case where the PUT and DELETE entry for a blob ended up in the same index
+    // segment after compaction (the DELETE wasn't eligible to be "counted"). The tests that exercise this path will
+    // need to find alternate ways of checking.
+    List<LogEntry> logEntriesInOrder = new ArrayList<>();
+    NavigableMap<Offset, IndexSegment> indexSegments = state.index.getIndexSegments();
+    for (String logSegmentName : logSegmentsUnderConsideration) {
+      LogSegment logSegment = state.log.getSegment(logSegmentName);
+      Offset indexSegmentStartOffset = new Offset(logSegmentName, logSegment.getStartOffset());
+      while (indexSegmentStartOffset != null && indexSegmentStartOffset.getName().equals(logSegmentName)) {
+        IndexSegment indexSegment = indexSegments.get(indexSegmentStartOffset);
+        List<MessageInfo> infos = new ArrayList<>();
+        indexSegment.getEntriesSince(null, new FindEntriesCondition(Long.MAX_VALUE), infos, new AtomicLong(0));
+        List<IndexEntry> indexEntries = new ArrayList<>();
+        for (MessageInfo info : infos) {
+          IndexValue value = indexSegment.find(info.getStoreKey());
+          indexEntries.add(new IndexEntry(info.getStoreKey(), value));
+        }
+        addToLogEntriesInOrder(indexEntries, logEntriesInOrder);
+        indexSegmentStartOffset = indexSegments.higherKey(indexSegmentStartOffset);
+      }
+    }
+    return logEntriesInOrder;
+  }
+
+  /**
+   * Adds {@link LogEntry} instances for all the {@code indexEntries} to {@code logEntriesInOrder}.
+   * @param indexEntries the index entries to process.
+   * @param logEntriesInOrder the list of {@link LogEntry} instances to add to.
+   */
+  private void addToLogEntriesInOrder(List<IndexEntry> indexEntries, List<LogEntry> logEntriesInOrder) {
+    Collections.sort(indexEntries, PersistentIndex.INDEX_ENTRIES_OFFSET_COMPARATOR);
+    for (IndexEntry entry : indexEntries) {
+      MockId id = (MockId) entry.getKey();
+      PersistentIndex.IndexEntryType entryType =
+          entry.getValue().isFlagSet(IndexValue.Flags.Delete_Index) ? PersistentIndex.IndexEntryType.DELETE
+              : PersistentIndex.IndexEntryType.PUT;
+      logEntriesInOrder.add(new LogEntry(id, entryType));
+    }
+  }
+
+  /**
+   * Checks the record of an id by ensuring that the {@link IndexValue} received from the store matches what is
+   * expected. Also checks the data if the key is not deleted and hard delete enabled.
+   * @param id the {@link MockId} whose record needs to be checked.
+   * @param options the {@link BlobReadOptions} received from the {@link PersistentIndex}.
+   * @throws IOException
+   */
+  private void checkRecord(MockId id, BlobReadOptions options) throws IOException {
+    List<BlobReadOptions> optionsList = new ArrayList<>();
+    optionsList.add(options);
+    MessageReadSet readSet = new StoreMessageReadSet(optionsList);
```

Comment:
Used `Arrays.asList()` because the constructor of `StoreMessageReadSet` does a `Collections.sort()` which will fail on unmodifiable lists (which `Collections.singletonList(options)` is).